test/test_broker.hpp

```diff
@@ -1787,15 +1903,47 @@ class test_broker {
         con_sp_t con;
         MQTT_NS::buffer client_id;
 
-        MQTT_NS::optional<MQTT_NS::will> will;
         MQTT_NS::optional<std::chrono::steady_clock::duration> will_delay;
         MQTT_NS::optional<std::chrono::steady_clock::duration> session_expiry_interval;
         std::shared_ptr<as::steady_timer> tim_session_expiry;
         MQTT_NS::optional<MQTT_NS::topic_alias_recv_map_t> topic_alias_recv;
+
         mi_inflight_message inflight_messages;
-        std::deque<offline_message> offline_messages;
         std::set<packet_id_t> qos2_publish_processed;
+
+        mi_offline_message offline_messages;
+
         std::set<sub_con_map::handle> handles; // to efficient remove
+
+        void update_will(as::io_context& ioc, MQTT_NS::optional<MQTT_NS::will> will, MQTT_NS::optional<std::chrono::steady_clock::duration> will_expiry_interval) {
+            tim_will_expiry.reset();
+            will_value = MQTT_NS::force_move(will);
+
+            if(will_value && will_expiry_interval) {
+                tim_will_expiry = std::make_shared<as::steady_timer>(ioc, will_expiry_interval.value());
+                tim_will_expiry->async_wait(
+                    [&](MQTT_NS::error_code ec) {
```

Comment:
Please don't use `[&]` capture with asynchronous callback.
I just added the coding rule. See https://github.com/redboltz/mqtt_cpp/wiki/Coding-Rules#details-of-lambda-expression

Reply:
Another comment.
Please capture

```cpp
[wp = std::weak_ptr<as::steady_timer>(tim_will_expiry)]
```

And use it as follows:

```cpp
if (auto sp = wp.lock()) {
    ....
}
```

Reply:
Here I actually use reset_will on shared_ptr<session_state> , but I can assume shared_ptr<session_state> still exists if tim_will_expiry shared_ptr still exist, correct ? tim_will_expiry is a member of session_state, timer will get cancelled when shared_ptr<session_state> is destroyed.

Reply:
> Here I actually use reset_will on shared_ptr<session_state> , but I can assume shared_ptr<session_state> still exists if tim_will_expiry shared_ptr still exist, correct ?

Yes.

 > tim_will_expiry is a member of session_state, timer will get cancelled when shared_ptr<session_state> is destroyed.

Yes, but as I mentioned several times, the following scenario could happen.

1. session_state starts destroying.
2. The member variables of session_state starts destroying.
3. tim_will_expiry is fired. Boost.Asio remove timer entry and post timer fired event (handler) to ioc.
4. tim_will_expiry starts destroying.
5. tim_will_expiry timer.cancel() called internally but **no timer entry matched**.
6. destroy other members and finish session_state destoroy.
7. timer handler is invoked. And access deallocated session_state. It's bad.